eclipse-theia/theia
# electron: enable context isolation (with inversify)

#### What it does

This is work on top of https://github.com/eclipse-theia/theia/pull/12299 to use Inversify in the preload context.

It also breaks up the monolithic `TheiaCoreAPI` into several smaller components.

The challenges that had to be solved are the following:

- How to share Inversify bindings from one context to another:

  An instance of `TheiaPreloadContext` is exposed to the browser global scope. When initializing the browser Inversify containers, we call `theiaPreloadContext.getAllServices()` to get pairs of `[serviceIdentifier: string, serviceProxy: object]` which we use to rebind each component as `bind(serviceIdentifier).toConstantValue(serviceProxy)`.

- How to share proxies of instances with prototypes:

  Electron only supports proxying objects used as records/dictionaries and won't follow prototype chains. This means we have to pre-process objects to make it work with Electron's cross-context APIs. This is done with the help of a decorator-based API `@proxyable()` and `@proxy()`. Then `IpcHandleConverter` is handling the pre-processing of instances using the metadata defined using the decorators.

It is now possible to bind and rebind preload components as a framework user.

Please see `./packages/core/ELECTRON-COMMON.md` for more details about how to contribute preload APIs.

I also write more information in my commit messages.

#### How to test

Execute commands such as:

- Toggle Developer Tools
- View: Toggle Full Screen
- File: Reveal in File Explorer

Maximazing/minimizing/closing windows should still work.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
